custom/zipline/models.py

```diff
@@ -156,6 +179,166 @@ class EmergencyOrder(models.Model):
         related_name='+', null=True)
 
 
+class EmergencyOrderPackage(models.Model):
+
+    class Meta:
+        index_together = [
+            ['order', 'package_number'],
+        ]
+
+        unique_together = [
+            ['order', 'package_number'],
+        ]
+
+    # (order, package_number) matches up with the same-named fields on EmergencyOrderStatusUpdate
+    order = models.ForeignKey('EmergencyOrder')
+    package_number = models.IntegerField()
+
+    # Same format as EmergencyOrder.products_requested; represents products and quantities
+    # in this package
+    products = jsonfield.JSONField(default=dict)
+
+    # Should either be STATUS_DISPATCHED, STATUS_DELIVERED, or STATUS_CANCELLED_IN_FLIGHT
+    status = models.CharField(max_length=126, default=EmergencyOrderStatusUpdate.STATUS_DISPATCHED)
+
+    # A pointer to the EmergencyOrderStatusUpdate representing the dispatched status update
+    # for this package
+    dispatched_status = models.ForeignKey('EmergencyOrderStatusUpdate', on_delete=models.PROTECT,
+        related_name='+', null=True)
+
+    # A pointer to the EmergencyOrderStatusUpdate representing the delivered status update
+    # for this package
+    delivered_status = models.ForeignKey('EmergencyOrderStatusUpdate', on_delete=models.PROTECT,
+        related_name='+', null=True)
+
+    # A pointer to the EmergencyOrderStatusUpdate representing the cancelled in flight status update
+    # for this package
+    cancelled_status = models.ForeignKey('EmergencyOrderStatusUpdate', on_delete=models.PROTECT,
+        related_name='+', null=True)
+
+    # The total cost of all products in this package, using info from OrderableProduct
+    cost = models.DecimalField(max_digits=12, decimal_places=3, null=True)
+
+    # The total weight of this package in grams, using info from OrderableProduct
+    weight = models.DecimalField(max_digits=12, decimal_places=3, null=True)
+
+    def update_calculated_fields(self):
+        """
+        Updates the cost and the weight of the package.
+        Note that this uses the active record for the product, so this method
+        should only be called in real-time (not on historical data).
+        """
+        from custom.zipline.api import log_zipline_exception
+
+        cost = Decimal(0)
+        weight = Decimal(0)
+
+        for code, data in self.products.iteritems():
+            try:
+                product = OrderableProduct.objects.get(domain=self.order.domain, code=code)
+            except OrderableProduct.DoesNotExist:
+                log_zipline_exception(
```

Comment:
don't know any better solution, but sending to the 500 emails doesn't feel quite right. this almost seems like an application error